src/truncate.jl

```diff
@@ -1,7 +1,41 @@
+"""
+    truncated(d, l, u):
+
+Truncate a distribution between `l` and `u`.
+Builds the most appropriate distribution for the type of `d`,
+the fallback is constructing a `Truncated` distribution.
+
+To implement a specialized truncated form for a distribution `D`,
+the method `truncate(d::D, l::T, u::T) where {T <: Real}`
+should be implemented.
+
+# Arguments
+- `d::UnivariateDistribution`: The original distribution.
+- `l::Real`: The lower bound of the truncation, which can be a finite value or `-Inf`.
+- `u::Real`: The upper bound of the truncation, which can be a finite value of `Inf`.
+
+Throws an error if `l >= u`.
+"""
+function truncated(d::UnivariateDistribution, l::Real, u::Real)
+    return truncated(d, promote(l, u)...)
+end
+
+function truncated(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+    l < u || error("lower bound should be less than upper bound.")
+    T2 = promote_type(T, eltype(d))
+    lcdf = isinf(l) ? zero(T2) : T2(cdf(d, l))
+    ucdf = isinf(u) ? one(T2) : T2(cdf(d, u))
+    tp = ucdf - lcdf
+    Truncated(d, promote(l, u, lcdf, ucdf, tp, log(tp))...)
+end
+
+truncated(d::UnivariateDistribution, l::Integer, u::Integer) = truncated(d, float(l), float(u))
+
 """
     Truncated(d, l, u):
 
 Construct a truncated distribution.
```

Comment:
Not `construct`, rather create a generic wrapper.

Reply:
done thanks